### DISCUSSION TITLE
Go to definition jumps to header instead, even with `compile_commands.json`

### DISCUSSION THREAD

**Discussion by znxuz** [clangd/clangd#1991] [General]
Maybe related to this [issue](https://github.com/clangd/clangd/issues/1592), but somehow it just won't jump to source file directly, even though it's on Linux in my case.

Directory structure:
```sh
.
├── build
│   ├── cmake
│   ├── CMakeCache.txt
│   ├── CMakeFiles
│   ├── cmake_install.cmake
│   ├── compile_commands.json
│   └── Makefile
├── CMakeLists.txt
├── Core
│   ├── Inc
│   │   ├── main.h
│   └── Src
│       ├── main.c
├── Drivers
│   └── STM32F7xx_HAL_Driver
│       ├── Inc
│       │   ├── stm32f7xx_hal.h
│           ├── ...
│       └── Src
│           ├── stm32f7xx_hal.c
│           ├── ...

```

In the `Core/Src/main.c` file, when trying to go to definition on a function defined in `Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal.c`, it only jumps to its declaration in the header `Drivers/STM32F7xx_HAL_Driver/Inc/stm32f7xx_hal.h`. Only after I manually open the source file, it will jump to it directly.

`compile_commands.json` generated by `cmake`:
```
...
{
  "directory": "/home/z/code/stm32f767zi/build",
  "command": "/usr/bin/arm-none-eabi-gcc -DDEBUG -DSTM32F767xx -DUSE_HAL_DRIVER -I/home/z/code/stm32f767zi/cmake/stm32cubemx/../../Core/Inc -I/home/z/code/stm32f767zi/cmake/stm32cubemx/../../Drivers/STM32F7xx_HAL_Driver/Inc -I/home/z/code/stm32f767zi/cmake/stm32cubemx/../../Drivers/STM32F7xx_HAL_Driver/Inc/Legacy -I/home/z/code/stm32f767zi/cmake/stm32cubemx/../../Drivers/CMSIS/Device/ST/STM32F7xx/Include -I/home/z/code/stm32f767zi/cmake/stm32cubemx/../../Drivers/CMSIS/Include  -mcpu=cortex-m7 -mfpu=fpv5-d16 -mfloat-abi=hard  -Wall -Wextra -Wpedantic -fdata-sections -ffunction-sections -O0 -g3 -g -std=gnu11 -o CMakeFiles/stm32f767zi.dir/Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal.c.obj -c /home/z/code/stm32f767zi/Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal.c",
  "file": "/home/z/code/stm32f767zi/Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal.c",
  "output": "CMakeFiles/stm32f767zi.dir/Drivers/STM32F7xx_HAL_Driver/Src/stm32f7xx_hal.c.obj"
},

...
```

Is `compile_commands.json` generated correctly here?

Server run via `clangd --background-index --log=error --query-driver=/usr/bin/arm-none-eabi-gcc'`

**Comment by HighCommander4** [comment]
Does the directory `build/.cache/clangd/index` exist? Does it contain a file named `stm32f7xx_hal.c.<hash>.idx`?

**Reply by znxuz** [reply]
I was just going to write that I couldn't find the `.cache` directory, but after rebuilding it the last time after deleting the `build`, the directory just magically appeared in the project root... And it gets re-generated every time after deleting it and restarting the client.

So weird that it didn't get generated at all by the server beforehand, but it contains the indexed file you mentioned and the whole thing works perfectly now.
